File: client/python/conductor/ConductorWorker.py
#
#  Copyright 2017 Netflix, Inc.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
from __future__ import print_function, absolute_import
import sys
import time
import subprocess
import conductor
from conductor.conductor import WFClientMgr
from threading import Thread
import socket

hostname = socket.gethostname()

class ConductorWorker:
    def __init__(self, server_url, thread_count, polling_interval):
        wfcMgr = WFClientMgr(server_url)
        self.workflowClient = wfcMgr.workflowClient
        self.taskClient = wfcMgr.taskClient
        self.thread_count = thread_count
        self.polling_interval = polling_interval

    def execute(self, task, exec_function):
        try:
            resp = exec_function(task)
            if(resp == None):
                raise Exception('Task execution function MUST return a response as a dict with status and output fields')
            task['status'] = resp['status']
            task['outputData'] = resp['output']
            task['logs'] = resp['logs']
            self.taskClient.updateTask(task)
        except Exception as err:
            print('Error executing task: ' + str(err))
            task['status'] = 'FAILED'
            self.taskClient.updateTask(task)

    def poll_and_execute(self, taskType, exec_function):
        while True:
            time.sleep(float(self.polling_interval))
            polled = self.taskClient.pollForTask(taskType, hostname)
            if(polled != None):
                self.execute(polled, exec_function)

    def start(self, taskType, exec_function, wait):
        print('Polling for task ' + taskType + ' at a ' + str(self.polling_interval) + ' ms interval with ' + str(self.thread_count) + ' threads for task execution, with worker id as ' + hostname)
        for x in range(0, int(self.thread_count)):
            thread = Thread(target=self.poll_and_execute, args=(taskType, exec_function, ))
            thread.daemon = True
            thread.start()
        if wait:
            while 1:
                time.sleep(1)

def exc(taskType, inputData, startTime, retryCount, status, callbackAfterSeconds, pollCount):
    print('Executing the function')
    return {'status': 'COMPLETED', 'output': {}}

def main():
    cc = ConductorWorker('http://localhost:8080/api', 5, 0.1)
    cc.start(sys.argv[1], exc, False)
    cc.start(sys.argv[2], exc, True)

if __name__ == '__main__':
    main()

File: polyglot-clients/go/conductorworker.go
// Copyright 2017 Netflix, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package conductor

import (
	"github.com/netflix/conductor/client/go/task"
	"log"
	"os"
	"time"
)

var (
	hostname, hostnameError = os.Hostname()
)

func init() {
	if hostnameError != nil {
		log.Fatal("Could not get hostname")
	}
}

type ConductorWorker struct {
	ConductorHttpClient *ConductorHttpClient
	ThreadCount         int
	PollingInterval     int
}

func NewConductorWorker(baseUrl string, threadCount int, pollingInterval int) *ConductorWorker {
	conductorWorker := new(ConductorWorker)
	conductorWorker.ThreadCount = threadCount
	conductorWorker.PollingInterval = pollingInterval
	conductorHttpClient := NewConductorHttpClient(baseUrl)
	conductorWorker.ConductorHttpClient = conductorHttpClient
	return conductorWorker
}

func (c *ConductorWorker) Execute(t *task.Task, executeFunction func(t *task.Task) (*task.TaskResult, error)) {
	taskResult, err := executeFunction(t)
	if taskResult == nil {
		log.Println("TaskResult cannot be nil: ", t.TaskId)
		return
	}
	if err != nil {
		log.Println("Error Executing task:", err.Error())
		taskResult.Status = task.FAILED
		taskResult.ReasonForIncompletion = err.Error()
	}

	taskResultJsonString, err := taskResult.ToJSONString()
	if err != nil {
		log.Println(err.Error())
		log.Println("Error Forming TaskResult JSON body")
		return
	}
	_, _ = c.ConductorHttpClient.UpdateTask(taskResultJsonString)
}

func (c *ConductorWorker) PollAndExecute(taskType string, domain string, executeFunction func(t *task.Task) (*task.TaskResult, error)) {
	for {
		time.Sleep(time.Duration(c.PollingInterval) * time.Millisecond)

		// Poll for Task taskType
		polled, err := c.ConductorHttpClient.PollForTask(taskType, hostname, domain)
		if err != nil {
			log.Println("Error Polling task:", err.Error())
			continue
		}
		if polled == "" {
			log.Println("No task found for:", taskType)
			continue
		}

		// Parse Http response into Task
		parsedTask, err := task.ParseTask(polled)
		if err != nil {
			log.Println("Error Parsing task:", err.Error())
			continue
		}

		// Found a task, so we send an Ack
		_, ackErr := c.ConductorHttpClient.AckTask(parsedTask.TaskId, hostname, domain)
		if ackErr != nil {
			log.Println("Error Acking task:", ackErr.Error())
			continue
		}

		// Execute given function
		c.Execute(parsedTask, executeFunction)
	}
}

func (c *ConductorWorker) Start(taskType string, domain string, executeFunction func(t *task.Task) (*task.TaskResult, error), wait bool) {
	log.Println("Polling for task:", taskType, "with a:", c.PollingInterval, "(ms) polling interval with", c.ThreadCount, "goroutines for task execution, with workerid as", hostname)
	for i := 1; i <= c.ThreadCount; i++ {
		go c.PollAndExecute(taskType, domain, executeFunction)
	}

	// wait infinitely while the go routines are running
	if wait {
		select {}
	}
}
